unit_tests/amp_test.cc

```diff
@@ -0,0 +1,96 @@
+#include "../include/amp/amp.hh"
+#include <gtest/gtest.h>
+
+// This is a unit test, magic numbers are called test data here
+// NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers)
+
+// TEST macro violates guidelines
+// NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-avoid-non-const-global-variables)
+TEST(AmpTest, EmptyReturn) {
+    const int scale = 1;
+    Audio novalue;
+
+    Audio out = Amplifier::amplify(novalue, scale);
+
+    Audio empty;
+    EXPECT_EQ(empty, out);
+}
+
+// TEST macro violates guidelines
+// NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-avoid-non-const-global-variables)
+TEST(AmpTest, Amplification) {
+    const int scale = 1;
+    const double scalehalf = 0.5;
+    const int scaledouble = 2;
+    Audio somevalues;
+    Audio somehalfvalues;
+    Audio somedoublevalues;
+    for (int i = 0; i < 100; i++) {
+        somevalues.push_back(i);
+        somehalfvalues.push_back(i / 2);
+        somedoublevalues.push_back(i * 2);
+    }
+    Audio out = Amplifier::amplify(somevalues, scale);
+    Audio outhalf = Amplifier::amplify(somevalues, scalehalf);
+    Audio outdouble = Amplifier::amplify(somevalues, scaledouble);
+
+    EXPECT_EQ(somevalues.size(), out.size());
+    EXPECT_EQ(somevalues, out);
+    EXPECT_EQ(somehalfvalues, outhalf);
+    EXPECT_EQ(somedoublevalues, outdouble);
+}
+
+// TEST macro violates guidelines
+// NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-avoid-non-const-global-variables)
+TEST(AmpTest, ThreadAndMessaging) {
+    Pipe<Audio> in_pipe;
+    Pipe<double> scale_pipe;
+    Pipe<Audio> out_pipe;
+
+    Amplifier a(in_pipe, scale_pipe, out_pipe, std::chrono::milliseconds(100));
+
+    int16_t single_sample = 2;
+    Audio amp_in;
+    amp_in.push_back(single_sample);
+    double scale_factor = 0.5;
+    Audio amp_out;
+    amp_out.push_back(single_sample * scale_factor);
+
+    a.run();
+
+    std::thread input_thread{[&in_pipe, &amp_in] {
+        auto *in = &in_pipe;
```

Comment:
i know some other tests use pointers, but references are better. should be `auto &in = in_pipe;`

Reply:
References are non mutable and do not require dereferencing when used, i.e. you can get rid of `in->queue.push(amp_in)` and replace it with `in.queue.push(amp_in)`

Reply:
Updated to use references.